flow/aggregates.py

```diff
@@ -206,30 +206,24 @@ def aggregator_function(jobs):
             for key, group in groupby(sorted(jobs, key=keyfunction), key=keyfunction):
                 yield group
 
-        return cls(aggregator_function, sort_by, reverse_order, select)
+        return cls(aggregator_function, sort_by, sort_ascending, select)
 
     def __eq__(self, other):
-        return type(self) == type(other) and \
-               not self._is_aggregate and not other._is_aggregate
+        return type(self) == type(other) and not self._is_aggregate and not other._is_aggregate
 
     def __hash__(self):
-        blob_l = [
-            hash(self._sort_by), hash(self._reverse_order), hash(self._is_aggregate),
-            self._get_unique_function_id(self._aggregator_function),
-            self._get_unique_function_id(self._select)
-        ]
-        blob = ','.join(str(attr) for attr in blob_l)
-
-        return int(_hash(blob), 16)
+        return hash((self._sort_by, self._sort_ascending, self._is_aggregate,
+                     self._get_unique_function_id(self._aggregator_function),
+                     self._get_unique_function_id(self._select)))
```

Comment:
We're severely over-complicating things here. Just take the hash of a tuple of hashable objects that uniquely define this instance. There's no need to do all the string operations and sha1 digests.
> The only required property is that objects which compare equal have the same hash value; it is advised to mix together the hash values of the components of the object that also play a part in comparison of objects by packing them into a tuple and hashing the tuple.
https://docs.python.org/3/reference/datamodel.html#object.__hash__

Reply:
Does the `_get_unique_function_id` even need to exist? @kidrahahjo why were you hashing just the bytecode of the function (which is what `_get_unique_function_id` returns for functions) rather than just hashing the function directly? I honestly don't remember thinking through whether it was necessary, I think I just assumed there was a reason you bothered to define that rather than just hashing the function directly and moved on. Any object with a `__code__` attribute should already be hashable though, right?

Reply:
@vyasr Since `aggregator_function` is defined internally hence the hash would depend on it's location.
I believe this was the main reason for not creating a hash of a method directly.

Reply:
This comment is just for others' reference. In my understanding, `_get_unique_function_id` is useful for the same reason we use `obj.__code__.co_code` for the "tags" on `_condition` objects in flow. It allows us to know that two function objects are the same if their bytecode is equivalent (even if they're not actually located in the same object). Hashing the bytecode (when possible) is a helpful way to uniquely identify the aggregator/select function.

Demonstration of how hashes of functions are not equal (even though the functions are equivalent), but hashing the bytecode shows their equivalence:
```python
>>> def foo():
...   print('asdf')
...
>>> def bar():
...   print('asdf')
...
>>>
>>> foo is bar
False
>>> hash(foo)
8793327702568
>>> hash(bar)
8793327702577
>>> foo.__code__.co_code
b't\x00d\x01\x83\x01\x01\x00d\x00S\x00'
>>> bar.__code__.co_code
b't\x00d\x01\x83\x01\x01\x00d\x00S\x00'
>>> assert hash(foo.__code__.co_code) == hash(bar.__code__.co_code)
```

Reply:
@kidrahahjo @bdice yes, that is true, but we're not using hashing to check equality or depending on being able to check equality at all any more since, as per our discussions, checking `co_code` is insufficient: for instance:

```
>>> def f(x):
...     return x**2
...
>>> def g(x):
...     return x**3
...
>>> f.__code__.co_code
b'|\x00d\x01\x13\x00S\x00'
>>> g.__code__.co_code
b'|\x00d\x01\x13\x00S\x00'
```

Although I pointed out some ways to get around the particular problem above (as well other even more subtle ways in which this comparison can fail), the point is that this id isn't really unique, right? So do we gain anything by defining a function whose name is not accurate and whose primary purpose is no longer valid? If `Aggregate.__eq__` only returns `True` if (roughly) `a is b`, then wouldn't it make sense for our hashes to also be distinct?

Reply:
@vyasr @kidrahahjo Wow, you're right. I didn't realize that would give different results. The 2 and 3 are stored in `__code__.co_consts`, and otherwise the functions' code looks identical. We should probably change this behavior to just use the behavior of `a is b` and not try to out-smart it.

Reply:
Right. Actually the situation even gets worse than that, I discussed some cases where comparing `__code__.co_consts` is also not sufficient (and possible ways to get around those edge cases as well). However, overall I think there are enough such possibilities that it's safest to avoid the problem altogether and just require strict object equality, which is also what we changed `__eq__` to do. @kidrahahjo could you make the relevant change here (on a separate branch/PR at this point)? Basically just remove the usage of this helper function and directly hash the object unless this will cause any other problems. @b-butler feel free to comment if you think I'm missing something here, but I think this is directly in line with my suggestion for `__eq__`.

Reply:
@bdice @vyasr The `__hash__` method does not have to distinguish all elements that are not equivalent it just requires that `a == b -> hash(a) == hash(b)`. We can do this for now, but I think eventually having a hash method that can distinguish between commonly used aggregates would be useful. hhttps://docs.python.org/3/reference/datamodel.html#object.__hash__

I don't think there is any safety concern given since if `a is b` then `a == b` and `hash(a) == hash(b)` in our model. `__hash__` is not necessarily meant in Python to provide a unique id. It is more for hashable collections like `dict` and `set`. We can safely give objects the same hash when they are not equal as long as our `__eq__` gives no false positives which the `is` operator ensures.

Reply:
I agree, in general hashes don't need to be equivalent and we only depend on `__eq__`. I just figured that for now we should keep the two roughly in sync.

signac at present basically assumes no hash collisions... but that's another story.

Reply:
To be clear, hash collisions in signac would be caught and raise errors. We have tests for this behavior. (Just didn’t want readers of the above comment to fear unexpected behavior.)

Reply:
Yes, hash collisions where two statepoints refer to the same job ID will raise an error. Nothing will break.

AFAIK hash collisions will raise the same `DestinationExistsError` as is raised when actually creating two identical jobs though. I'm not aware of any logic that explicitly checks for the _reason_ and indicates to the user that there's a hash collision, so the error will look the same as if the user did something like

```
j1 = project.open_job(dict(a=1)).init()
j2 = project.open_job(dict(a=1, b=2)).init()
del j2.sp.b  # This will raise a DestinationExistsError because now j2 and j1 have the same statepoint.
```

In the event that a user somehow managed to trigger a collision (which is obviously highly unlikely, and not something that I think we need to deal with), the resulting error message would prevent any corruption of the data space, it would just be very confusing. That's all I was referring to, nothing needs to be done there.